BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs

```diff
@@ -169,10 +173,84 @@ internal virtual void Enqueue(Transmission transmission)
 
             if (!this.Storage.Enqueue(transmissionGetter))
             {
+                transmission.HasFlushTask = false;
                 TelemetryChannelEventSource.Log.TransmitterStorageSkipped(transmission.Id);
             }
         }
 
+        internal Task<bool> FlushAsync(Transmission transmission, CancellationToken cancellationToken)
+        {
+            TaskStatus taskStatus = TaskStatus.Canceled;
+            if (!cancellationToken.IsCancellationRequested)
+            {
+                transmission.HasFlushTask = true;
+                this.Enqueue(transmission);
+
+                try
+                {
+                    this.Storage.IncrementFlushAsyncCounter();
+                    taskStatus = this.MoveTransmissionsAndWaitForSender(transmission.FlushAsyncId, cancellationToken);
+                }
+                catch (Exception exp)
+                {
+                    taskStatus = TaskStatus.Faulted;
+                    TelemetryChannelEventSource.Log.TransmissionFlushAsyncWarning(exp.ToString());
+                }
+                finally
+                {
+                    this.Storage.DecrementFlushAsyncCounter();
+                }
+            }
+
+            Task<bool> flushTaskStatus = null;
+            if (taskStatus == TaskStatus.Canceled)
+            {
+                flushTaskStatus = TaskEx.FromCanceled<bool>(cancellationToken);
+            }
+            else if (taskStatus == TaskStatus.RanToCompletion && transmission.HasFlushTask)
+            {
+                flushTaskStatus = this.successTask;
+            }
+            else
+            {
+                flushTaskStatus = this.failedTask;
+            }
+
+            return flushTaskStatus;
+        }
+
+        internal Task<bool> MoveTransmissionsAndWaitForSender(CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return TaskEx.FromCanceled<bool>(cancellationToken);
+            }
+
+            var isStorageEnqueueSuccess = MoveTransmissions(this.Buffer.Dequeue, this.Storage.Enqueue, this.Buffer.Size);  
+            TelemetryChannelEventSource.Log.MovedFromBufferToStorage();
+            var senderStatus = this.Sender.WaitForPreviousTransmissionsToComplete(cancellationToken).ConfigureAwait(false).GetAwaiter().GetResult();
+            if (senderStatus == TaskStatus.Canceled)
+            {
+                return TaskEx.FromCanceled<bool>(cancellationToken);
+            }
+
+            return senderStatus == TaskStatus.RanToCompletion && isStorageEnqueueSuccess ? this.successTask : this.failedTask; 
+        }
+
+        internal TaskStatus MoveTransmissionsAndWaitForSender(long transmissionFlushAsyncId, CancellationToken cancellationToken)
```

Comment:
these 2 methods can be merged/made to call other, to avoid copy-pasting.

Reply:
Return type of both of these methods are different and they call `WaitForPreviousTransmissionsToComplete' method with different parameters. 

Reason for me to split these functions are to avoid handling TaskCanceledException in the `Transmitter.FlushAsync` call. Felt handling exception is costlier here.